oxidation/libparsec/crates/serialization_format/src/lib.rs

```diff
@@ -1,65 +1,118 @@
+    let pathname = parse_macro_input!(path as LitStr).value();
+    let path = path_from_str(&pathname);
+    let family = path
+        .file_name()
+        .and_then(|os_str| os_str.to_str())
+        .expect("Invalid filename");
+    let collection = if path.is_dir() {
+        let content = content_from_dir(&path).expect("Failed to get content from directory");
+        parser::ProtocolCollection::with_protocols(family, content)
+    } else {
+        let content = content_from_file(&path).expect("Failed to get content from file");
+        parser::ProtocolCollection::with_protocol(family, content)
+    };
+    let collection = intermediate::ProtocolCollection::from(collection);
+    collection.quote().into_token_stream().into()
+}
+
+pub(crate) fn path_from_str(path: &str) -> PathBuf {
+    let manifest_dir_path: PathBuf = std::env::var("CARGO_MANIFEST_DIR")
         .expect("CARGO_MANIFEST_DIR should be set")
-        .parse::<PathBuf>()
-        .expect("CARGO_MANIFEST_DIR must be a valid path");
+        .parse()
+        .expect("CARGO_MANIFEST_DIR should be a valid path");
     manifest_dir_path.join(path)
 }
 
-fn content_from_file(path: &PathBuf) -> String {
-    let file = File::open(path).expect("Cannot open the json file");
+pub(crate) fn content_from_dir(path: &PathBuf) -> anyhow::Result<Vec<parser::Protocol>> {
+    let dir = std::fs::read_dir(path)
+        .map_err(anyhow::Error::new)
+        .context(format!("Reading directory `{}`", path.to_string_lossy()))?;
+    dir.filter_map(|entry| entry.ok())
+        .map(|entry| content_from_file(&entry.path()))
+        .collect()
+}
+
+pub(crate) fn content_from_file(path: &PathBuf) -> anyhow::Result<parser::Protocol> {
+    let filename = path.to_string_lossy();
+    let file = File::open(path)
+        .map_err(anyhow::Error::new)
+        .context("Opening file")?;
     let buf = BufReader::new(file);
     let mut content = String::new();
     for (i, line) in buf.lines().enumerate() {
-        let line = line.unwrap_or_else(|_| panic!("Non-Utf-8 character found in line {i}"));
-        let line = match line.split_once("//") {
-            Some((line, _)) => line,
-            None => &line,
-        };
+        let line =
+            line.unwrap_or_else(|_| panic!("{}:{} Non UTF-8 characters found", &filename, i));
+        let line = line
+            .split_once("//")
+            .map(|(before, _after)| before)
+            .unwrap_or(&line);
         content.push_str(line)
     }
-    content
+    content_from_str(&content, &filename)
 }
 
-fn content_from_dir(path: &PathBuf) -> String {
-    let dir = std::fs::read_dir(path).expect("Cannot read the directory");
-    dir.filter_map(|entry| entry.ok())
-        .map(|entry| content_from_file(&entry.path()))
-        .collect::<Vec<_>>()
-        .join(",")
+pub(crate) fn content_from_str(
+    content: &str,
+    origin: &str,
+) -> Result<parser::Protocol, anyhow::Error> {
+    serde_json::from_str(content)
```

Comment:
why use serde_json instead of `miniserde::json` ?

Reply:
Continue this discussion in https://github.com/Scille/parsec-cloud/pull/3076#discussion_r983460451